File: src/telemetry/pollWorker.js
/**
 * Background Poll Worker
 * Polls Graph Engine API and writes metrics to InfluxDB
 */

const graphEngineClient = require('../clients/graphEngineClient');
const InfluxWriter = require('../clients/influxWriter');
const config = require('../config/config');

class PollWorker {
  constructor() {
    this.influxWriter = new InfluxWriter();
    this.intervalId = null;
    this.isRunning = false;
    this.polling = false;
    this.lastPollAt = null;
    this.lastSuccessAt = null;
  }

  /**
   * Start the poll worker
   */
  start() {
    if (!config.telemetryWorker.enabled) {
      console.log('[PollWorker] Disabled (TELEMETRY_WORKER_ENABLED=false)');
      return;
    }

    if (this.isRunning) {
      console.warn('[PollWorker] Already running');
      return;
    }

    console.log(`[PollWorker] Starting with ${config.telemetryWorker.pollIntervalMs}ms interval`);
    this.isRunning = true;

    // Run immediately on start
    this.poll();

    // Schedule recurring polls
    this.intervalId = setInterval(() => {
      this.poll();
    }, config.telemetryWorker.pollIntervalMs);
  }

  /**
   * Stop the poll worker
   */
  async stop() {
    if (!this.isRunning) {
      return;
    }

    console.log('[PollWorker] Stopping...');
    this.isRunning = false;

    if (this.intervalId) {
      clearInterval(this.intervalId);
      this.intervalId = null;
    }

    await this.influxWriter.close();
    console.log('[PollWorker] Stopped');
  }

  /**
   * Execute one poll cycle
   */
  async poll() {
    // Overlap protection - skip if previous poll still running
    if (this.polling) {
      console.warn('[PollWorker] Previous poll still running, skipping this cycle');
      return;
    }

    this.polling = true;
    this.lastPollAt = new Date();

    try {
      console.log('[PollWorker] Polling Graph Engine...');

      // 1. Fetch Request/Latency Metrics (via Snapshot)
      let services = [];
      let edges = [];

      try {
        const snapshotResult = await graphEngineClient.getMetricsSnapshot();

        if (snapshotResult.ok && snapshotResult.data) {
          // Transform Graph Engine schema to InfluxDB schema
          services = (snapshotResult.data.services || []).map(svc => {
            const hasTraffic = svc.rps && svc.rps > 0;
            return {
              name: svc.name,
              namespace: svc.namespace,
              requestRate: svc.rps ?? null,
              errorRate: hasTraffic ? svc.errorRate : null,
              p50: null,
              p95: hasTraffic ? svc.p95 : null,
              p99: null,
              availability: null
            };
          });

          edges = (snapshotResult.data.edges || []).map(edge => {
            const hasTraffic = edge.rps && edge.rps > 0;
            return {
              from: edge.from,
              to: edge.to,
              namespace: edge.namespace,
              requestRate: edge.rps ?? null,
              errorRate: hasTraffic ? edge.errorRate : null,
              p50: null,
              p95: hasTraffic ? edge.p95 : null,
              p99: null
            };
          });
        }
      } catch (err) {
        console.error(`[PollWorker] Snapshot fetch failed: ${err.message}`);
      }

      // 2. Fetch Infrastructure Metrics (via Services with Placement)
      // This is a separate call because snapshot is optimized for light edges
      let infraData = { nodes: [], services: [] };
      try {
        const servicesResult = await graphEngineClient.getServicesWithPlacement();
        if (servicesResult.ok && servicesResult.data && servicesResult.data.services) {
          // Extract Nodes from the services list (Graph Engine returns services -> placement -> nodes)
          // We need to de-duplicate nodes since multiple services run on same nodes
          const nodeMap = new Map();

          servicesResult.data.services.forEach(svc => {
            if (svc.placement && svc.placement.nodes) {
              svc.placement.nodes.forEach(node => {
                if (!node.node) return;
                if (!nodeMap.has(node.node)) {
                  nodeMap.set(node.node, node);
                } else {
                  // Merge pods
                  const existing = nodeMap.get(node.node);
                  if (node.pods && node.pods.length > 0) {
                    // Add pods that aren't already listed (simple check by name)
                    node.pods.forEach(p => {
                      if (!existing.pods.some(ep => ep.name === p.name)) {
                        existing.pods.push(p);
                      }
                    });
                  }
                }
              });
            }
          });

          infraData.nodes = Array.from(nodeMap.values());
        }
      } catch (err) {
        console.error(`[PollWorker] Infra fetch failed: ${err.message}`);
      }

      // 3. Write to InfluxDB
      const promises = [];

      if (services.length > 0) {
        promises.push(this.influxWriter.writeServiceMetrics(services));
      }

      if (edges.length > 0) {
        promises.push(this.influxWriter.writeEdgeMetrics(edges));
      }

      if (infraData.nodes.length > 0) {
        promises.push(this.influxWriter.writeInfrastructureMetrics(infraData));
      }

      await Promise.all(promises);

      this.lastSuccessAt = new Date();
      console.log(`[PollWorker] Poll complete: ${services.length} services, ${edges.length} edges, ${infraData.nodes.length} nodes`);

    } catch (error) {
      console.error(`[PollWorker] Poll failed: ${error.message}`);
      // Continue running despite errors
    } finally {
      this.polling = false;
    }
  }
}

// Singleton instance
let workerInstance = null;

/**
 * Get or create the singleton poll worker instance
 */
function getWorker() {
  if (!workerInstance) {
    workerInstance = new PollWorker();
  }
  return workerInstance;
}

module.exports = {
  getWorker,
  PollWorker
};

File: pkg/worker/pollworker.go
package worker

import (
	"context"
	"log"
	"sync"
	"time"

	"predictive-analysis-engine/pkg/clients/graph"
	"predictive-analysis-engine/pkg/clients/telemetry"
	"predictive-analysis-engine/pkg/config"
)

type PollWorker struct {
	graphClient     *graph.Client
	telemetryClient *telemetry.TelemetryClient
	cfg             *config.Config
	stopCh          chan struct{}
	wg              sync.WaitGroup
	running         bool
	runLock         sync.Mutex
}

func NewPollWorker(cfg *config.Config, gClient *graph.Client, tClient *telemetry.TelemetryClient) *PollWorker {
	return &PollWorker{
		graphClient:     gClient,
		telemetryClient: tClient,
		cfg:             cfg,
		stopCh:          make(chan struct{}),
	}
}

func (w *PollWorker) Start() {
	if !w.cfg.TelemetryWorker.Enabled {
		log.Println("[PollWorker] Disabled (TELEMETRY_WORKER_ENABLED=false)")
		return
	}

	w.runLock.Lock()
	if w.running {
		w.runLock.Unlock()
		log.Println("[PollWorker] Already running")
		return
	}
	w.running = true
	w.runLock.Unlock()

	log.Printf("[PollWorker] Starting with %dms interval\n", w.cfg.TelemetryWorker.PollIntervalMs)

	w.wg.Add(1)
	go func() {
		defer w.wg.Done()
		w.poll()

		ticker := time.NewTicker(time.Duration(w.cfg.TelemetryWorker.PollIntervalMs) * time.Millisecond)
		defer ticker.Stop()

		for {
			select {
			case <-w.stopCh:
				return
			case <-ticker.C:
				w.poll()
			}
		}
	}()
}

func (w *PollWorker) Stop() {
	w.runLock.Lock()
	if !w.running {
		w.runLock.Unlock()
		return
	}
	w.running = false
	w.runLock.Unlock()

	log.Println("[PollWorker] Stopping...")
	close(w.stopCh)
	w.wg.Wait()

	log.Println("[PollWorker] Stopped")
}

func (w *PollWorker) poll() {
	log.Println("[PollWorker] Polling Graph Engine...")
	ctx := context.Background()

	var servicePoints []telemetry.ServicePoint
	var edgePoints []telemetry.EdgePoint

	snapshot, err := w.graphClient.GetMetricsSnapshot(ctx)
	if err != nil {
		log.Printf("[PollWorker] Snapshot fetch failed: %v\n", err)
	} else if snapshot != nil {

		for _, svc := range snapshot.Services {
			hasTraffic := svc.RPS > 0

			var rps, errRate, p95, p50, p99, avail *float64

			r := svc.RPS
			rps = &r

			if hasTraffic {
				e := svc.ErrorRate
				errRate = &e
				p95Val := svc.P95
				p95 = &p95Val
			}

			servicePoints = append(servicePoints, telemetry.ServicePoint{
				Name:         svc.Name,
				Namespace:    svc.Namespace,
				RequestRate:  rps,
				ErrorRate:    errRate,
				P95:          p95,
				P50:          p50,
				P99:          p99,
				Availability: avail,
			})
		}

		for _, edge := range snapshot.Edges {
			hasTraffic := edge.RPS > 0
			var rps, errRate, p95, p50, p99 *float64

			r := edge.RPS
			rps = &r

			if hasTraffic {
				e := edge.ErrorRate
				errRate = &e
				p := edge.P95
				p95 = &p
			}

			edgePoints = append(edgePoints, telemetry.EdgePoint{
				From:        edge.From,
				To:          edge.To,
				Namespace:   edge.Namespace,
				RequestRate: rps,
				ErrorRate:   errRate,
				P95:         p95,
				P50:         p50,
				P99:         p99,
			})
		}
	}

	var nodePoints []telemetry.PkgNodePoint
	var podPoints []telemetry.PkgPodPoint

	services, err := w.graphClient.GetServices(ctx)
	if err != nil {
		log.Printf("[PollWorker] Infra fetch failed: %v\n", err)
	} else {

		type uniqueNode struct {
			NodePlacement graph.NodePlacement
			Pods          []graph.PodInfo
		}
		uniqueNodes := make(map[string]*uniqueNode)

		for _, svc := range services {

			for _, node := range svc.Placement.Nodes {
				if node.Node == "" {
					continue
				}

				if _, exists := uniqueNodes[node.Node]; !exists {

					podsCopy := make([]graph.PodInfo, len(node.Pods))
					copy(podsCopy, node.Pods)
					uniqueNodes[node.Node] = &uniqueNode{
						NodePlacement: node,
						Pods:          podsCopy,
					}
				} else {

					existing := uniqueNodes[node.Node]
					for _, newPod := range node.Pods {
						found := false
						for _, exPod := range existing.Pods {
							if exPod.Name == newPod.Name {
								found = true
								break
							}
						}
						if !found {
							existing.Pods = append(existing.Pods, newPod)
						}
					}
				}
			}
		}

		for _, u := range uniqueNodes {

			cpuUse := u.NodePlacement.Resources.CPU.UsagePercent
			cpuCores := float64(u.NodePlacement.Resources.CPU.Cores)
			ramUsed := float64(u.NodePlacement.Resources.RAM.UsedMB)
			ramTotal := float64(u.NodePlacement.Resources.RAM.TotalMB)
			podCount := float64(len(u.Pods))

			nodePoints = append(nodePoints, telemetry.PkgNodePoint{
				Name:            u.NodePlacement.Node,
				CPUUsagePercent: &cpuUse,
				CPUTotalCores:   &cpuCores,
				RAMUsedMB:       &ramUsed,
				RAMTotalMB:      &ramTotal,
				PodCount:        &podCount,
			})

			for _, pod := range u.Pods {
				ram := pod.RAMUsedMB
				cpuPct := pod.CPUUsagePercent

				podPoints = append(podPoints, telemetry.PkgPodPoint{
					Name:            pod.Name,
					NodeName:        u.NodePlacement.Node,
					RAMUsedMB:       &ram,
					CPUUsagePercent: &cpuPct,
				})
			}
		}
	}

	if len(servicePoints) > 0 {
		if err := w.telemetryClient.WriteServiceMetrics(ctx, servicePoints); err != nil {
			log.Printf("[PollWorker] Write service metrics failed: %v", err)
		}
	}

	if len(edgePoints) > 0 {
		if err := w.telemetryClient.WriteEdgeMetrics(ctx, edgePoints); err != nil {
			log.Printf("[PollWorker] Write edge metrics failed: %v", err)
		}
	}

	if len(nodePoints) > 0 {

		if err := w.telemetryClient.WriteInfrastructureMetrics(ctx, nodePoints, podPoints); err != nil {
			log.Printf("[PollWorker] Write infra metrics failed: %v", err)
		}
	}

	log.Printf("[PollWorker] Poll complete: %d services, %d edges, %d nodes\n", len(servicePoints), len(edgePoints), len(nodePoints))
}
